fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraSessionImpl.java

```diff
@@ -0,0 +1,205 @@
+
+    // The default timeout is 3 minutes
+    public static final String DEFAULT_TIMEOUT = Long.toString(ofMinutes(3).toMillis());
+
+    public static final String TIMEOUT_SYSTEM_PROPERTY = "fcrepo.session.timeout";
+
+    private final Session jcrSession;
+    private final String id;
+    private final Instant created;
+    private final ConcurrentHashMap<String, String> sessionData;
+    private Instant expires;
+
+    /**
+     * A key for looking up the transaction id in a session key-value pair
+     */
+    public static final String FCREPO_TX_ID = "fcrepo.tx.id";
+
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    /**
+     * Create a Fedora session with a JCR session and a username
+     * @param session the JCR session
+     */
+    public FedoraSessionImpl(final Session session) {
+        this.jcrSession = session;
+
+        created = now();
+        id = randomUUID().toString();
+        expires = created.plus(operationTimeout());
+        sessionData = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void commit() {
+        try {
+            if (jcrSession.isLive()) {
+                final ObservationManager obs = jcrSession.getWorkspace().getObservationManager();
+                final ObjectNode json = mapper.createObjectNode();
+                sessionData.forEach(json::put);
+                obs.setUserData(mapper.writeValueAsString(json));
+                jcrSession.save();
+            }
+        } catch (final javax.jcr.AccessDeniedException ex) {
+            throw new AccessDeniedException(ex);
+        } catch (final RepositoryException | JsonProcessingException ex) {
+            throw new RepositoryRuntimeException(ex);
+        }
+    }
+
+    @Override
+    public void expire() {
+        expires = now();
+        try {
+            if (jcrSession.isLive()) {
+                jcrSession.refresh(false);
+                jcrSession.logout();
+            }
+        } catch (final RepositoryException ex) {
+            throw new RepositoryRuntimeException(ex);
+        }
+    }
+
+    @Override
+    public Instant updateExpiry(final Duration amountToAdd) {
+        if (jcrSession.isLive()) {
+            expires = now().plus(amountToAdd);
+        }
+        return expires;
+    }
+
+    @Override
+    public Instant getCreated() {
+        return created;
+    }
+
+    @Override
+    public Optional<Instant> getExpires() {
+        return of(expires);
+    }
+
+    @Override
+    public String getId() {
+        return id;
+    }
+
+    @Override
+    public String getUserId() {
+        return jcrSession.getUserID();
+    }
+
+    @Override
+    public Map<String, String> getNamespaces() {
+        return NamespaceTools.getNamespaces(jcrSession);
+    }
+
+    /**
+     *  Add session data
+     *  @param key the data key
+     *  @param value the data value
+     *
+     *  Note: while the FedoraSession inteface permits multi-valued
```

Comment:
Have you verified that the current codebase does not require appending values to an existing UserData key?